dropshot/src/error.rs

```diff
@@ -290,15 +304,118 @@ impl HttpError {
             error_code,
             internal_message,
             external_message,
+            headers: None,
         }
     }
 
+    /// Mutably borrow the `http::HeaderMap` associated with this error, if one
+    /// exists.
+    pub fn headers_mut(&mut self) -> Option<&mut http::HeaderMap> {
+        self.headers.as_deref_mut()
+    }
+
+    /// Borrow the `http::HeaderMap` associated with this error, if one
+    /// exists.
+    pub fn headers_ref(&self) -> Option<&http::HeaderMap> {
+        self.headers.as_deref()
+    }
+
+    /// Adds a header to the [`http::HeaderMap`] of headers to add to responses
+    /// generated from this error.
+    ///
+    /// If this error does not already have a header map (`self.header_map` is
+    /// `None`), this method creates one.
+    ///
+    /// # Returns
+    /// - [`Ok`]`(&mut Self)` if the provided `name` and `value` are a valid
+    ///   header name and value, respectively.
+    /// - [`Err`]`(`[`http::Error`]`)` if the header name or value is invalid,
+    ///   or the `HeaderMap` is full.
+    pub fn add_header<K, V>(
+        &mut self,
+        name: K,
+        value: V,
+    ) -> Result<&mut Self, http::Error>
+    where
+        http::HeaderName: TryFrom<K>,
+        <http::HeaderName as TryFrom<K>>::Error: Into<http::Error>,
+        http::HeaderValue: TryFrom<V>,
+        <http::HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
+    {
+        let name = <http::HeaderName as TryFrom<K>>::try_from(name)
+            .map_err(Into::into)?;
+        let value = <http::HeaderValue as TryFrom<V>>::try_from(value)
+            .map_err(Into::into)?;
+        self.headers
+            .get_or_insert_with(|| Box::new(http::HeaderMap::default()))
+            .try_append(name, value)?;
+        Ok(self)
+    }
+
+    /// Adds a header to the [`http::HeaderMap`] of headers to add to responses
+    /// generated from this error, taking the error by value.
+    ///
+    /// If this error does not already have a header map (`self.header_map` is
+    /// `None`), this method creates one.
+    ///
+    /// Unlike [`HttpError::set_header`], this method takes `self` by value,
+    /// allowing it to be chained to form an expression that returns an
+    /// `HttpError`. However, because this takes `self` by value, returning an
+    /// error for an invalid header name or value will discard the `HttpError`.
+    /// To avoid this, use [`HttpError::set_header`] instead.
```

Comment:
This is a bit awkward. Another option could be to make the method take `HeaderName` and `HeaderValue` only, instead of types that can be converted into them; then, it could be infallible, instead of potentially eating the `HttpError`.

I did it this way because it's similar to the `http::Response::builder()` API, and allows users to just pass strings as the header name/value, rather than having to convert them in user code. But, returning a `Result` and maybe throwing away the error is also unfortunate. I could be convinced to change this --- what do y'all think?